client/app/services/widget.js

```diff
@@ -106,29 +106,32 @@ function WidgetFactory($http, Query, Visualization, dashboardGridOptions) {
     }
 
     load(force, maxAge) {
-      this.loading = true;
-      this.refreshStartedAt = moment();
-
       if (!this.visualization) {
         return undefined;
       }
 
-      if (force || this.queryResult === undefined) {
+      // Both `this.data` and `this.queryResult` are query result objects;
+      // `this.data` is last loaded query result;
+      // `this.queryResult` is currently loading query result;
+      // while widget is refreshing, `this.data` !== `this.queryResult`
+
+      if (force || (this.queryResult === undefined)) {
         if (maxAge === undefined || force) {
           maxAge = force ? 0 : undefined;
         }
         this.queryResult = this.getQuery().getQueryResult(maxAge);
+        this.loading = true;
+        this.refreshStartedAt = moment();
```

Comment:
Shouldn't this come before the line where we call `getQueryResult`?

Reply:
It doesn't matter, actually. But if you think it will look better - I can do it

Reply:
@arikfr Done

Reply:
It's not about looks :) It's about correctness. Currently the code here assumes something about `getQueryResult`'s implementation -- that it will return immediately. But it's not necessarily will always be the case.